File: topics/mpi/solutions/03.MPI_Coll/fortran/2.scatter_solution.f90
PROGRAM scatter

!==============================================================!
!                                                              !
! This file has been written as a sample solution to an        !
! exercise in a course given at the CSCS Summer School.        !
! It is made freely available with the understanding that      !
! every copy of this file must include this header and that    !
! CSCS take no responsibility for the use of the enclosed      !
! teaching material.                                           !
!                                                              !
! Purpose: a simple scatter                                    !
!                                                              !
! Contents: F-Source                                           !
!==============================================================!


   USE MPI
   IMPLICIT NONE
   INTEGER i, rank, size, senddata(20), receivedata, ierror

   CALL MPI_Init(ierror)
   CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierror)
   CALL MPI_Comm_size(MPI_COMM_WORLD, size, ierror)

   IF (size.GT.20) THEN
      IF (rank.EQ.0) THEN
         WRITE (*,*) "do not use more than 20 processors"
         CALL MPI_Finalize(ierror)
      END IF
   END IF
   IF (rank.EQ.0) THEN
      DO i=1, size, 1
         WRITE (6,*) 'enter a value:'
         CALL FLUSH(6)
         READ (*,*) senddata(i)
      END DO
   END IF

   ! scatter the value of senddata of rank 0 to receivedata of all ranks
   CALL MPI_Scatter(senddata, 1, MPI_INTEGER, receivedata, 1, MPI_INTEGER, 0, MPI_COMM_WORLD, ierror)

   WRITE (*,*) "I am rank", rank, "and the value is", receivedata

   CALL MPI_Finalize(ierror)

END PROGRAM

File: topics/mpi/solutions/03.MPI_Coll/cxx/2.scatter_solution.c
/****************************************************************
 *                                                              *
 * This file has been written as a sample solution to an        *
 * exercise in a course given at the CSCS Summer School.        *
 * It is made freely available with the understanding that      *
 * every copy of this file must include this header and that    *
 * CSCS take no responsibility for the use of the enclosed      *
 * teaching material.                                           *
 *                                                              *
 * Purpose: Simple scatter                                      *
 *                                                              *
 * Contents: C-Source                                           *
 *                                                              *
 ****************************************************************/


#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

int main(int argc, char *argv[]) {
   int i, rank, size, senddata[20], receivedata;
   MPI_Init(&argc, &argv);
   MPI_Comm_rank(MPI_COMM_WORLD, &rank);
   MPI_Comm_size(MPI_COMM_WORLD, &size);
   if (size>20){
      if (rank==0)
         printf("do not use more than 20 processors\n");
      MPI_Finalize();
      exit(1);
   }
   if (rank==0){
      for (i=0; i<size; i++){
         printf("enter a value:\n");
         fflush(stdout);
         scanf ("%d",&senddata[i]);
      }
   }

   /* scatter the value of senddata of rank 0 to receivedata of all ranks */
   MPI_Scatter(&senddata, 1, MPI_INT, &receivedata, 1, MPI_INT, 0, MPI_COMM_WORLD);

   printf("I am rank %i and the value is %i\n", rank, receivedata);
   MPI_Finalize();
   return 0;
}
